### PR TITLE
refactor: Improve types and error handling

### PR DESCRIPTION
**Pull request by KShivendu**
Improve types and error handling

### CODE REVIEW

File: agentai/api.py [hunk 1]
```diff
@@ -1,31 +1,36 @@
 """
 API functions for the agentai package
 """
-from typing import Callable
+from typing import Callable, List
```
Comment (NirantK) [inline agentai/api.py:4]: prefer `list`, `dict` over `List`, `Dict` in Python>=3.9

File: agentai/api.py [hunk 2]
```diff
@@ -1,31 +1,36 @@
...
+from .conversation import Conversation, Message
 from .openai_function import ToolRegistry
 
 logger.disable(__name__)
 
+class InvalidOutput(Exception):
+    pass
 
-@retry(retry=retry_if_exception_type(ValueError), stop=stop_after_attempt(3))
+retry_on_exceptions = retry_if_exception_type((InvalidOutput, requests.HTTPError, KeyError))
```
Comment (NirantK) [inline agentai/api.py:19]: SQLErrors?

Comment (KShivendu) [reply agentai/api.py:19]: Hmm. Then maybe we need to find a way to retry on anything other than ValueError.

File: agentai/openai_function.py [hunk 3]
```diff
@@ -114,14 +114,17 @@ def dec(obj):
 
     return dec
 
+class Tool(Callable[..., Any]):
+    json_info: str
+
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply agentai/openai_function.py:119]: Lol. I think the last time I saw it, it was a string. Will update.

File: agentai/api.py (hunk 2 above)
Comment (KShivendu) [reply agentai/api.py]: Done.

File: agentai/openai_function.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (KShivendu) [reply agentai/openai_function.py]: Done.

File: agentai/api.py (hunk 1 above)
Comment (KShivendu) [reply agentai/api.py]: Done.

File: agentai/api.py [hunk 4]
```diff
@@ -37,16 +42,17 @@ def chat_complete(
...
         json=json_data,
     )
+    response.raise_for_status()
     if return_function_params:
         message = response.json()["choices"][0]
         if message["finish_reason"] == "function_call":
             return response
         else:
-            raise ValueError(f"Unexpected message: {message}")
+            raise Exception(f"Unexpected message: {message}")
```
Comment (NirantK) [inline agentai/api.py:51]: Too broad, add something specific e.g. ValueError or similar? Help the dev understand why this is wrong in someway

File: agentai/api.py [hunk 5]
```diff
@@ -66,10 +72,10 @@ def get_function_arguments(message, conversation: Conversation, function_registr
                 message, conversation, function_registry=function_registry, model=model
             )
         return function_arguments
-    raise ValueError(f"Unexpected message: {message}")
+    raise Exception(f"Unexpected message: {message}")
 
 
-@retry(retry=retry_if_exception_type(ValueError), stop=stop_after_attempt(3))
+@retry_unless_invalid
```
Comment (NirantK) [inline agentai/api.py:78]: Ohh neat, I didn't know about this!

File: agentai/api.py [hunk 6]
```diff
@@ -6,26 +6,31 @@
...
+from .conversation import Conversation, Message
 from .openai_function import ToolRegistry
 
 logger.disable(__name__)
 
+class InvalidInputError(Exception):
+    pass
 
-@retry(retry=retry_if_exception_type(ValueError), stop=stop_after_attempt(3))
+retry_unless_invalid = retry(retry=retry_unless_exception_type(InvalidInputError), stop=stop_after_attempt(3))
```
Comment (NirantK) [inline agentai/api.py:19]: Don't hard code the `stop_after_attempt` here

**Review by NirantK — CHANGES_REQUESTED** [review]

File: agentai/api.py (hunk 5 above)
Comment (NirantK) [inline agentai/api.py:78]: Keep the retry decorators separate, we'll have different backoff plans for one pass vs 2 pass functions

**Review by NirantK — CHANGES_REQUESTED** [review]

File: agentai/api.py (hunk 4 above)
Comment (KShivendu) [reply agentai/api.py]: Done. Using `VaueError`

File: agentai/api.py (hunk 6 above)
Comment (KShivendu) [reply agentai/api.py]: FIxed.

File: agentai/api.py (hunk 5 above)
Comment (KShivendu) [reply agentai/api.py]: Done.

**Review by NirantK — APPROVED** [review]

**Pull request merged**